### ISSUE TITLE
[nll] hash borrows in scope for better performance

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#53159]
So I have an idea for how to improve performance in cases like html5ever where there are tons of borrows in scope at certain points. Right now, we have a pretty naive algorithm that implements over all borrows in scope:

https://github.com/rust-lang/rust/blob/18925dee25ce649562d203e72068e3a57b60b153/src/librustc_mir/borrow_check/path_utils.rs#L40-L52

I think we would do better if we stored the "borrows in scope" organized by the "root local" of the borrow:

https://github.com/rust-lang/rust/blob/18925dee25ce649562d203e72068e3a57b60b153/src/librustc_mir/borrow_check/place_ext.rs#L20-L23

In fact, I think we could do a "quick hack" to make this work. The `BorrowSet` already has an index of borrow indices based on their "root locals":

https://github.com/rust-lang/rust/blob/18925dee25ce649562d203e72068e3a57b60b153/src/librustc_mir/borrow_check/borrow_set.rs#L44-L45

So what we can do is to modify `each_borrow_involving_path`. Instead of taking an iterator over borrows in scope, it would take some sort of `is_borrow_in_scope: impl Fn(BorrowIndex) -> bool` function. It would then find the root local (if any) of the access path, find all borrows that apply to that root local, and iterate over those, testing if each of them is in scope. 

We could make this more efficient in many ways -- e.g., using bitsets and things -- but this basic change should already eliminate some of the horrible `O(n^2)` behavior.

**Comment by nikomatsakis** [comment]
This should be a big win for html5ever -- `each_borrow_involving_path` currently stands at 41% there!

**Comment by nikomatsakis** [comment]
This will likely not help html5ever as much once https://github.com/rust-lang/rust/pull/53177 lands.

https://github.com/rust-lang/rust/issues/53189  is also relevant, as it will also remove many uses of `each_borrow_involving_path`.

**Comment by nikomatsakis** [comment]
Zulip topic: https://rust-lang.zulipchat.com/#narrow/stream/122657-wg-nll/topic/.2353159.20hash.20borrows.20in.20scope

**Comment by Mark-Simulacrum** [comment]
Moving to RC 2 -- not a blocker.

**Comment by nikomatsakis** [comment]
Removing from all milestones as this just doesn't release like a Rust 2018 blocker (perf is "good enough"). Marked as NLL-deferred.

**Comment by pnkfelix** [comment]
NLL premeeting triage. Unclear what priority to assign here, since it is not clear whether or not this would still be much of a win. (Its possible it might be; I just don't know.) Continuing to leave off a P-label; hopfeully I'll revisit next week with more information.

**Comment by nikomatsakis** [comment]
I'm inclined to just close this issue, personally, although I guess the idea may still be actionable and relevant.

**Comment by pnkfelix** [comment]
NLL triage: closing. If we later decide that this idea has merit, we can reopen (or open a fresh issue, or just implement a PR).
